src/jdk.jdeps/share/classes/com/sun/tools/jnativescan/ClassResolver.java

```diff
@@ -0,0 +1,176 @@
+                    ClassDesc desc =  model.thisClass().asSymbol();
+                    classMap.put(desc, new Info(m.moduleName(), m.path(), model));
+                } catch (IOException e) {
+                    throw new RuntimeException(e);
+                }
+            });
+        }
+        return new ScannedModuleClassResolver(loaded, classMap);
+    }
+
+    static ClassResolver forSystemModules(Runtime.Version version) {
+        String platformName = String.valueOf(version.feature());
+        PlatformProvider platformProvider = ServiceLoader.load(PlatformProvider.class).findFirst().orElseThrow();
+        PlatformDescription platform;
+        try {
+            platform = platformProvider.getPlatform(platformName, null);
+        } catch (PlatformProvider.PlatformNotSupported e) {
+            throw new JNativeScanFatalError("Release: " + platformName + " not supported", e);
+        }
+        JavaFileManager fm = platform.getFileManager();
+        return new SystemModuleClassResolver(fm);
+    }
+
+    record Info(String moduleName, Path jarPath, ClassModel model) {}
+
+    public abstract void forEach(BiConsumer<ClassDesc, ClassResolver.Info> action);
+    public abstract Optional<ClassResolver.Info> lookup(ClassDesc desc);
+
+    @Override
+    public abstract void close() throws IOException;
+
+    private static class ScannedModuleClassResolver extends ClassResolver {
+
+        private final List<JarFile> jars;
+        private final Map<ClassDesc, ClassResolver.Info> classMap;
+
+        public ScannedModuleClassResolver(List<JarFile> jars, Map<ClassDesc, Info> classMap) {
+            this.jars = jars;
+            this.classMap = classMap;
+        }
+
+        public void forEach(BiConsumer<ClassDesc, ClassResolver.Info> action) {
+            classMap.forEach(action);
+        }
+
+        public Optional<ClassResolver.Info> lookup(ClassDesc desc) {
+            return Optional.ofNullable(classMap.get(desc));
+        }
+
+        @Override
+        public void close() throws IOException {
+            for (JarFile jarFile : jars) {
+                jarFile.close();
+            }
+        }
+    }
+
+    private static class SystemModuleClassResolver extends ClassResolver {
+
+        private final JavaFileManager platformFileManager;
+        private final Map<String, String> packageToSystemModule;
+        private final Map<ClassDesc, Info> cache = new HashMap<>();
+
+        public SystemModuleClassResolver(JavaFileManager platformFileManager) {
+            this.platformFileManager = platformFileManager;
+            this.packageToSystemModule = packageToSystemModule(platformFileManager);
+        }
+
+        private static Map<String, String> packageToSystemModule(JavaFileManager platformFileManager) {
+            try {
+                Set<JavaFileManager.Location> locations = platformFileManager.listLocationsForModules(
```

Comment:
FWIW: +1 on using the modules from the `ct.sym` rather than runtime modules here!